File: python/backtracking/subsets.py
class Solution:
    # @return a ListNode
    def __init__(self):
        self.result = []

    def subsets(self, S):
        self.dfs(0, S, [])
        return self.result

    def dfs(self, start, S, valuelist):
        valuelist.sort()
        self.result.append(valuelist)
        for i in xrange(start, len(S)):
            self.dfs(i+1, S, valuelist+[S[i]]) 

if __name__ == "__main__":
    solution = Solution()

    print solution.subsets([1, 2, 3])

File: java/Subsets.java
import java.util.*;
public class Solution {
    public List<List<Integer>> result = new ArrayList<List<Integer>>();
    public List<List<Integer>> subsets(int[] S) {
        Arrays.sort(S);
        List<Integer> valuelist = new ArrayList<Integer>();
        this.dfs(S, valuelist);
        return result;
    }

    public void dfs(int[] S, List<Integer> valuelist) {
        this.result.add(valuelist);
        if (S.length == 0) {
            return;
        }
        for (int i = 0; i < S.length; i++) {
            ArrayList<Integer> newList = new ArrayList<Integer>(valuelist);
            newList.add(S[i]);
            dfs(Arrays.copyOfRange(S, i+1, S.length), newList);
        }
    }
}
